### PR TITLE
feature/improved-unit-parsing

### PR DESCRIPTION
**Pull request by builtbywill**

### CODE REVIEW

**Comment by builtbywill** [comment]
I just found another unit parsing bug related to prefixes. Since I have this branch open still, I'll plan to fix it here, but feel free to review the rest of the branch.

File: tests/all_good_test.py [hunk 1]
```diff
@@ -173,9 +173,11 @@ class TestAllGood(object):
         ("\\exponentialE ^{\\imaginaryI \\pi }", exp(_Mul(I, pi), evaluate=False)),
         ("\\exponentialE ^\\pi ", exp(pi, evaluate=False)),
         ("\\int x\\differentialD x", Integral(x, x)),
-        ("\\int x\\differentialD \\theta ", Integral(x, theta)),
+        ("\\int \\differentialD \\theta x", Integral(x, theta)),
```
Comment (tragessere) [inline tests/all_good_test.py:176]: @builtbywill I'm a little confused by these added tests. The original version of this looks correct to me, not the new one. And the two added tests below seem to be wrong too. 

It's converting an integral that's differentiating with respect to two variables at the same time, which I'm not sure you can even do. The latex looks to me like it's trying to make the Sympy expression `Integral(1, _Mul(x, \\theta))`. Sympy complains that it's an invalid limit, and Wolfram doesn't like it either.

Comment (tragessere) [reply tests/all_good_test.py:176]: I'm assuming we have someone trying to do that since it's the name of the issue over in the solution engine?

Comment (gregstrasters) [reply tests/all_good_test.py:176]: @tragessere based on this comment in the solution engine MR, i think the plan is just to get latex2sympy to be able to parse it so we can throw the correct error?
image

Comment (tragessere) [reply tests/all_good_test.py:176]: Hmm, that feels like a weird reason to have a test that verifies it parses to the wrong thing.

Comment (builtbywill) [reply tests/all_good_test.py:176]: @tragessere @gregstrasters honestly I was under the impression that the order of the "body" of the integral didn't matter, but yeah that's apparently not the case 🤔 

I just did some sleuthing and the `\differentialD` changes did cause a change in how integrals are parsed. before that update, the expression `\\int dxy` _was_ able to parse but would result in `y*Integral(1, x)`, where anything after the "differential" symbol is _not_ part of the integral

So there's technically another issue here. I'll take a look at it.

TLDR: it is still a valid expression and should be parse-able, but the integral should end after the "dx"

Comment (builtbywill) [reply tests/all_good_test.py:176]: The main thing this branch was fixing was the issue where given `\differentialD xy` part, the "x" wouldn't be paired up with the "d" like it should be, since "xy" was getting joined together when it shouldn't

now letters are _never_ joined together except when explicitly using `process_sympy_as_unit`

**Review by tragessere — APPROVED** [review]

**Review by jkdufair — COMMENTED** [review]
@builtbywill I'm afraid I'm not exactly sure I understand these changes. Would you be up for a short walkthrough? Or a detailed explanation here maybe?

**Review by jkdufair — APPROVED** [review]
@builtbywill :thumps:!
